Sources/NIOSOCKS/Messages/ClientGreeting.swift

```diff
@@ -0,0 +1,57 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+struct ClientGreeting: Hashable {
+    let version: UInt8 = 5
+    var methods: [AuthenticationMethod]
+}
+
+extension ByteBuffer {
+    
+    mutating func readClientGreeting() throws -> ClientGreeting? {
+        return try self.parseUnwindingIfNeeded { buffer in
+            guard
+                try buffer.readAndValidateProtocolVersion() != nil,
+                let numMethods = buffer.readInteger(as: UInt8.self)
+            else {
+                return nil
+            }
+            
+            var methods: [AuthenticationMethod] = []
+            methods.reserveCapacity(Int(numMethods))
+            for _ in 0..<numMethods {
+                guard let method = buffer.readInteger(as: UInt8.self) else {
```

Comment:
Might be good to have a quick fast-exit path that `buffer.readableBytes >= numMethods` and then force-unwrap the `readInteger` call. This will avoid us being open to a quadratic loop if the user sends a lot of methods in a slowloris style.